### ISSUE TITLE
Issues when deleting users

### ISSUE THREAD

**Issue by jojoob** [academic-moodle-cooperation/moodle-mod_organizer#126]
We stumbled over some errors and warnings while we bulk deleted some users:

## wrong include
The PHP error stated that $CFG is not defined in `admin/user/lib.php`... Turns out that `organizer_prepare_and_send_message()` in [messaging.php](https://github.com/academic-moodle-cooperation/moodle-mod_organizer/blob/cc6e7ba53a62f688fe16797a1fd30821eac9b6f4/messaging.php#L241) wants to include a `lib.php` (probably mod/organizer/lib.php but since there is no path defined the lib.php inside the original script directory is included...). And since $CFG is not defined global in this function...

I think the include_once statement can safely removed since [in line 34 locallib.php is already included](https://github.com/academic-moodle-cooperation/moodle-mod_organizer/blob/cc6e7ba53a62f688fe16797a1fd30821eac9b6f4/messaging.php#L34) which in turn includes lib.php.

```
[Tue Nov 28 13:40:59.620777 2023] [php:warn] [pid 3229602] PHP Warning:  Undefined variable $CFG in /srv/moodle/admin/user/lib.php on line 3, referer: https://example.com/admin/user/user_bulk_delete.php
[Tue Nov 28 13:40:59.620806 2023] [php:warn] [pid 3229602] PHP Warning:  Attempt to read property "dirroot" on null in /srv/moodle/admin/user/lib.php on line 3, referer: https://example.com/admin/user/user_bulk_delete.php
[Tue Nov 28 13:40:59.620828 2023] [php:warn] [pid 3229602] PHP Warning:  require_once(/user/filters/lib.php): Failed to open stream: No such file or directory in /srv/moodle/admin/user/lib.php on line 3, referer: https://example.com/admin/user/user_bulk_delete.php
[Tue Nov 28 13:40:59.622081 2023] [php:notice] [pid 3229602] Default exception handler: Exception - Failed opening required '/user/filters/lib.php' (include_path='[dirroot]/lib/pear:.:/usr/share/php') Debug:
	Error code: generalexceptionmessage
	* line 3 of /admin/user/lib.php: Error thrown
	* line 241 of /mod/organizer/messaging.php: call to include_once()
	* line 88 of /mod/organizer/classes/observer.php: call to organizer_prepare_and_send_message()
	* line ? of unknownfile: call to mod_organizer_observer::user_enrolment_deleted()
	* line 155 of /lib/classes/event/manager.php: call to call_user_func()
	* line 75 of /lib/classes/event/manager.php: call to core\\event\\manager::process_buffers()
	* line 835 of /lib/classes/event/base.php: call to core\\event\\manager::dispatch()
	* line 2304 of /lib/enrollib.php: call to core\\event\\base->trigger()
	* line 2806 of /lib/enrollib.php: call to enrol_plugin->unenrol_user()
	* line 1143 of /lib/enrollib.php: call to enrol_plugin->user_delete()
	* line 4158 of /lib/moodlelib.php: call to enrol_user_delete()
	* line 32 of /admin/user/user_bulk_delete.php: call to delete_user()
, referer: https://example.com/admin/user/user_bulk_delete.php
```

## typo in constant variable name
**this one is already fixed (28a25ba51f20c74b4e26a461668e79f4133adea6)**

```
Default exception handler: Exception - Undefined constant "ORGANIZER_GROUPMODE_EXISTINGGROUP" Debug: 
	Error code: generalexceptionmessage
	* line 71 of /mod/organizer/classes/observer.php: Error thrown
	* line ? of unknownfile: call to mod_organizer_observer::user_enrolment_deleted()
	* line 155 of /lib/classes/event/manager.php: call to call_user_func()
	* line 75 of /lib/classes/event/manager.php: call to core\\event\\manager::process_buffers()
	* line 835 of /lib/classes/event/base.php: call to core\\event\\manager::dispatch()
	* line 2304 of /lib/enrollib.php: call to core\\event\\base->trigger()
	* line 2806 of /lib/enrollib.php: call to enrol_plugin->unenrol_user()
	* line 1143 of /lib/enrollib.php: call to enrol_plugin->user_delete()
	* line 4158 of /lib/moodlelib.php: call to enrol_user_delete()
	* line 33 of /admin/user/user_bulk_delete.php: call to delete_user()
, referer: https://example.com/admin/user/user_bulk_delete.php?
```

The constant's actual name has an 's' at the end: [ORGANIZER_GROUPMODE_EXISTINGGROUPS](https://github.com/academic-moodle-cooperation/moodle-mod_organizer/blob/cc6e7ba53a62f688fe16797a1fd30821eac9b6f4/lib.php#L42)

## Invalid database query parameter value
```
[Tue Nov 28 14:39:00.478597 2023] [php:notice] [pid 3229603] [client 141.45.244.2:47802] PHP Notice:  Exception encountered in event observer 'mod_organizer_observer::user_enrolment_deleted': Coding error detected, it must be fixed by a programmer: Invalid database query parameter value<ul style="text-align: left" data-rel="backtrace"><li>
line 902 of /lib/dml/moodle_database.php: call to moodle_database->detect_objects()</li><li>
line 1024 of /lib/dml/pgsql_native_moodle_database.php: call to moodle_database->fix_sql_params()</li><li>
line 1679 of /lib/dml/moodle_database.php: call to pgsql_native_moodle_database->get_records_sql()</li><li>
line 1651 of /lib/dml/moodle_database.php: call to moodle_database->get_record_sql()</li><li>
line 1630 of /lib/dml/moodle_database.php: call to moodle_database->get_record_select()</li><li>
line 321 of /mod/organizer/messaging.php: call to moodle_database->get_record()</li><li>
line 88 of /mod/organizer/classes/observer.php: call to organizer_prepare_and_send_message()</li><li>
line ? of unknownfile: call to mod_organizer_observer::user_enrolment_deleted()</li><li>
line 155 of /lib/classes/event/manager.php: call to call_user_func()</li><li>
line 75 of /lib/classes/event/manager.php: call to core\\event\\manager::process_buffers()</li><li>
line 835 of /lib/classes/event/base.php: call to core\\event\\manager::dispatch()</li><li>
line 2304 of /lib/enrollib.php: call to core\\event\\base->trigger()</li><li>
line 2806 of /lib/enrollib.php: call to enrol_plugin->unenrol_user()</li><li>
line 1143 of /lib/enrollib.php: call to enrol_plugin->user_delete()</li><li>
line 4158 of /lib/moodlelib.php: call to enrol_user_delete()</li><li>
line 33 of /admin/user/user_bulk_delete.php: call to delete_user()</li></ul> in /srv/moodle/lib/weblib.php on line 3351,
referer: https://example.com/admin/user/user_bulk_delete.php?
```

The `$data` variable, which is an object, is [feeded to the `$DB->get_record()` function](https://github.com/academic-moodle-cooperation/moodle-mod_organizer/blob/cc6e7ba53a62f688fe16797a1fd30821eac9b6f4/messaging.php#L323) as an integer (id). I guess `$data->id` was meant? The same applies to `'register_notify_teacher:queue'`, `'register_notify_teacher:register'` and `'group_registration_notify:student:register'` cases.

**Comment by jojoob** [comment]
The error "typo in constant variable name" is already fixed by 28a25ba51f20c74b4e26a461668e79f4133adea6.

**Comment by bluetom** [comment]
Yes, problem #2 ("typo in constant variable name") is fixed in the current version 4.2.2 of the organizer.
Problem #3 ("Invalid database query parameter value") is fixed in the next version 4.2.3 which is supposed to be published this week or next week the lastest.
I will add the fix of problem #3 at the eleventh hour and report here when the publication is done.
Thank you for reporting!

**Comment by bluetom** [comment]
The new organizer version 4.2.3 contains fixes for these problems: https://moodle.org/plugins/mod_organizer/versions

**Comment by jojoob** [comment]
That's good but we need a backport to a Moodle 4.1 compatible plugin version.

**Comment by jojoob** [comment]
Hi @bluetom, are you willing to publish an official version for Moodle 4.1 containing this fix? I would appreciate it to avoid a local backport. :)

**Comment by bluetom** [comment]
Hi,
yes, we are planning to publish a version 4.1 of the organizer as soon as the current work on version 4.3 is finished. The version 4.1 will propably  be published next week...

**Comment by jojoob** [comment]
I backported the fixes for the three bugs onto Version 4.1.2 but did not tested it yet: [MOODLE_401_BACKPORTS](https://github.com/rzhtwberlin/moodle-mod_organizer/tree/MOODLE_401_BACKPORTS) ([diff](https://github.com/academic-moodle-cooperation/moodle-mod_organizer/compare/MOODLE_401_STABLE...rzhtwberlin:moodle-mod_organizer:MOODLE_401_BACKPORTS))
